src/Razor/test/Microsoft.VisualStudio.LanguageServerClient.Razor.Test/HtmlCSharp/DefaultLSPProgressListenerTest.cs

```diff
@@ -0,0 +1,249 @@
+
+            var token = Guid.NewGuid().ToString();
+            var notificationTimeout = TimeSpan.FromSeconds(2);
+            var cts = new CancellationTokenSource();
+
+            var lspProgressListener = new DefaultLSPProgressListener(languageServiceBroker);
+
+            // Act
+            var listenerAdded = lspProgressListener.TryListenForProgress(
+                token,
+                onProgressNotifyAsync: async (value, ct) => { await Task.Delay(5); },
+                notificationTimeout,
+                cts.Token,
+                out var onCompleted);
+
+            // Assert
+            Assert.True(listenerAdded);
+            Assert.NotNull(onCompleted);
+            Assert.False(onCompleted.IsCompleted);
+        }
+
+        [Fact]
+        public void TryListenForProgress_DuplicateRegistration_ReturnsFalse()
+        {
+            // Arrange
+            var languageServiceBroker = Mock.Of<ILanguageServiceBroker2>();
+
+            var token = Guid.NewGuid().ToString();
+            var notificationTimeout = TimeSpan.FromSeconds(2);
+            var cts = new CancellationTokenSource();
+
+            var lspProgressListener = new DefaultLSPProgressListener(languageServiceBroker);
+
+            // Act
+            _ = lspProgressListener.TryListenForProgress(
+                token,
+                onProgressNotifyAsync: async (value, ct) => { await Task.Delay(5); },
+                notificationTimeout,
+                cts.Token,
+                out _);
+            var listenerAdded = lspProgressListener.TryListenForProgress(
+                token,
+                onProgressNotifyAsync: async (value, ct) => { await Task.Delay(5); },
+                notificationTimeout,
+                cts.Token,
+                out var onCompleted);
+
+            // Assert
+            Assert.False(listenerAdded);
+            Assert.Null(onCompleted);
+        }
+
+        [Fact]
+        public async Task TryListenForProgress_TaskNotificationTimeoutAfterNoInitialProgress()
+        {
+            // Arrange
+            var languageServiceBroker = Mock.Of<ILanguageServiceBroker2>();
+
+            var token = Guid.NewGuid().ToString();
+            var notificationTimeout = TimeSpan.FromSeconds(2);
+            var cts = new CancellationTokenSource();
+
+            var lspProgressListener = new DefaultLSPProgressListener(languageServiceBroker);
+
+            // Act 1
+            var listenerAdded = lspProgressListener.TryListenForProgress(
+                token,
+                onProgressNotifyAsync: async (value, ct) => { await Task.Delay(5); },
+                notificationTimeout,
+                cts.Token,
+                out var onCompleted);
+
+            // Assert 1
+            Assert.True(listenerAdded);
+            Assert.NotNull(onCompleted);
+            Assert.False(onCompleted.IsCompleted);
+
+            // Act 2
+            await Task.Delay(TimeSpan.FromSeconds(3));
```

Comment:
ooo this is super racey. For instance on some CI's even an operation that takes milliseconds can sometimes take seconds to complete. I'd recommend following a similar model the background document generator where you modify the implementation to have several reset events to allow you to control certain behaviors at test time: https://github.com/dotnet/aspnetcore-tooling/blob/7bf5b094bad355d4389be4ca0c4eb82343853b57/src/Razor/src/Microsoft.VisualStudio.Editor.Razor/BackgroundDocumentGenerator.cs#L66-L81

Reply:
Thanks for the link, I took a look through the reset events implementation. After giving the LSPProgressListener implementation some more thought, I'm not convinced we need to utilize reset events of this for, for it. 

The task delays were being used to verify that the appropriate time had passed (in a real non-CI situation). If we proceed to using reset events, the end validation wouldn't really give us any more information over just using the `await onCompleted` we already have implemented. This gives us the added benefit of further verifying that functionality as well. 

I've updated the implementation to use `await onCompleted`, and increased the timeouts where necessary, significantly such that CI specific race conditions shouldn't be an issue. (https://github.com/dotnet/aspnetcore-tooling/pull/2119/commits/1ef86b0ca3ace26a8cbbed14cafcd707e4ac08c6)

Please let me know if you still have concerns, I'll be sure to keep reset event based testing in mind for next time when I see potentially racey situations 😄

Reply:
> CI specific race conditions shouldn't be an issue.

Famous last words 😄 

That being said, lets give it a shot. i don't want to block on tests for a preview4 release and if it really doesn't end up causing CI issues then it's for all the best. Over time we'll be able to tell if they're CI resilient.

Thanks for taking a look!. Approving!

Reply:
> Famous last words 😄

It's going to fail on the merge commit isn't it. 👀